feat(#3742): Split `ParsingErrors` on Two Separate Classes

volodya-lombrozo opened the pull request:
In this PR I:
1. Createad a separate package `errors` for all error handlers and realted classes.
2. Split `ParsingErrors` on `EoParserErrors` and `ParsingErrors`
3. Added supplementary classes like `Lines` and `ErrorDirectives` to share common logic across all the new classes.

When this PR seems having a lot of lines, it just moves some parts of the code across the codebase.

Closes: #3742.

volodya-lombrozo commented:
@yegor256 Could you review these changes, please?

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/errors/ErrorDirectives.java:
```diff
@@ -0,0 +1,70 @@
+ * Error directives.
+ * @since 0.50
+ */
+final class ErrorDirectives implements Iterable<Directive> {
```
@volodya-lombrozo we already have `DrProgram` and `DrListing` classes. Maybe this one can follow the same convention?

volodya-lombrozo replied:
@yegor256 Renamed to `DrErrors`

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/errors/EoParserErrors.java:
```diff
@@ -36,22 +35,15 @@
+ * @since 0.50
  */
-final class ParsingErrors extends BaseErrorListener implements Iterable<Directive> {
+public final class EoParserErrors extends BaseErrorListener implements Iterable<Directive> {
```
@volodya-lombrozo same here, we have `Dr` prefix for class that are iterables of `Directive`. Maybe can use it too?

volodya-lombrozo replied:
@yegor256 Renamed to `DrEoParserErrors`

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/errors/Lines.java:
```diff
@@ -0,0 +1,66 @@
+     * @param number The line number.
+     * @return The line.
+     */
+    String line(final int number) {
```
@volodya-lombrozo isn't it the same as `lines.get(i).orElse("")`? Looks like this class is an overkill.

volodya-lombrozo replied:
@yegor256 In case of using `lines.get(i).orElse("")` - you will get NPE. 

Basically, I need to share the same logic between `DrEoParserErrors` and `DrParsingErrors` classes (the `String line(final int number)` method). I can't use `static` method by obvious reasons. Thus, I have left with only one option - to create a `Lines` object. Do you know any other way how to share the logic and avoid creating `Lines` class?

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/errors/ParsingErrors.java:
```diff
@@ -0,0 +1,114 @@
+ *
+ * @since 0.30.0
+ */
+public final class ParsingErrors extends BaseErrorListener implements Iterable<Directive> {
```
@volodya-lombrozo same here: if it's an iterable of `Directive`, the `Dr` prefix will be a good name pattern, I believe

yegor256 replied:
@volodya-lombrozo also, implementation inheritance is not what we like to use. Composition is a better choice usually. See https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html

volodya-lombrozo replied:
@yegor256 Renamed to `DrParsingErrors`.

volodya-lombrozo replied:
> @volodya-lombrozo also, implementation inheritance is not what we like to use. Composition is a better choice usually. See https://www.yegor256.com/2016/09/13/inheritance-is-procedural.html

I would like to avoid inheritance, but `BaseErrorListener` is an ANTLR class and it's the way how they handle errors.

yegor256 reviewed (changes requested):
@volodya-lombrozo looks interesting, a few comment above

volodya-lombrozo commented:
@yegor256 Could you have a look one more time, please?

yegor256 commented:
@volodya-lombrozo the naming of classes is rather confusing here: they look almost the same. Why don't you use the names suggested in the puzzle: `ParserErrors` and `LexerErrors`?

volodya-lombrozo commented:
> @volodya-lombrozo the naming of classes is rather confusing here: they look almost the same. Why don't you use the names suggested in the puzzle: `ParserErrors` and `LexerErrors`?

@yegor256 I agree that the naming here is confusing. Let me try to explain. Before that PR, we had only one error listener: `ParsingErrors`, and that's all. I thought it would be fine to split it into two separate parts—`LexerErrors` and `ParserErrors`. When I started implementing this PR, I noticed that you were using the same `ParsingErrors` for both `EoSyntax` and `PhiSyntax`, which works fine until we implement language-specific error messages. So, instead of `LexerErrors` and `ParserErrors`, I would have to create `EoLexerErrors`, `EoParserErrors`, `PhiLexerErrors`, and `PhiParserErrors`. That would result in too many classes (and I wouldn't be able to sell that idea to you! 😊). As a result, I decided to leave `ParsingErrors` as it is and simply add one more class: `DrEoParserErrors`, specifically for EO errors. Please feel free to ask questions if anything is unclear.

volodya-lombrozo commented:
@yegor256 Recent changes:

1. Renamed `ParsingErrors` -> `GeneralErrors` (?)
2. Removed `package`
3. `GeneralErrors` and `EoParserErrors` don't implement `Iterable<Directive>` anymore. They create directives through the `directive()` method instead.

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/GeneralErrors.java:
```diff
@@ -0,0 +1,115 @@
+        return this.errors.size();
+    }
+
+    /**
```
@volodya-lombrozo this method is here by mistake?

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/PhiSyntax.java:
```diff
@@ -107,7 +107,7 @@ public XML parsed() throws IOException {
             new XMLDocument(
                 new Xembler(
-                    new Directives(xel).append(spy).append(this.extra)
+                    new Directives(xel).append(spy.directives()).append(this.extra)
```
@volodya-lombrozo looks like it should be `.append(new DrErrors(spy))`, isn't it?

yegor256 reviewed (changes requested):
@volodya-lombrozo it's much better now! please, see my comments above

volodya-lombrozo commented:
@yegor256 Please review these changes: [fbbaa29](https://github.com/objectionary/eo/pull/3751/commits/fbbaa29e8b8c67eb40ea9862e3ea67a19dd578c1). I can accommodate your suggestions in this manner (it's the only way). If you don't like the solution, we can revert them.

yegor256 commented on eo-parser/src/main/java/org/eolang/parser/DrErrors.java:
```diff
@@ -0,0 +1,68 @@
+     * @param errors The errors.
+     */
+    DrErrors(final Errors errors) {
+        this.errors = errors;
```
@volodya-lombrozo here, you only use `errors.all()` method here, don't use `size()`. Why can't you use `Iterable` instead of `Errors` here?

yegor256 commented:
@volodya-lombrozo can't you simply calculate the size of `Iterable` with the help of `LengthOf` from Cactoos? This `Errors` class looks redundant here. It only complicates the design adding no value.

volodya-lombrozo commented:
@yegor256 Sure, please, check this one: https://github.com/objectionary/eo/pull/3751/commits/d4764a698da350c8338bd0f9eb5035c4f56d8c1a
I've done it. However, I don't prefer such a solution because we hide an abstraction, that we actually have. Although, I should admit, that we have less code now.

yegor256 reviewed (approved):
LGTM

yegor256 commented:
@volodya-lombrozo thanks!
